File: sdk/node-holons/src/relay.js
'use strict';

const path = require('node:path');

const observability = require('./observability');
const relayPb = require('./gen/relay/v1/relay_pb');
const relayGrpc = require('./gen/relay/v1/relay_grpc_pb');

class RelayService {
    constructor(options = {}) {
        this.downstream = options.downstream || options.downstreamConn || null;
        this.received = 0;
    }

    async tick(call, callback) {
        try {
            const req = call.request || new relayPb.TickRequest();
            this.received += 1;
            const count = this.received;
            const obs = observability.current();
            const slug = responderSlug(obs);
            const uid = obs.cfg.instanceUid || '';
            obs.logger('tick').info('tick received', {
                sender: req.getSender(),
                note: req.getNote(),
                responder_slug: slug,
                responder_uid: uid,
            });
            const counter = obs.counter(
                'cascade_ticks_total',
                'Ticks received by this cascade node.',
                { responder_uid: uid },
            );
            if (counter) counter.inc();

            const out = new relayPb.TickResponse();
            out.setResponderSlug(slug);
            out.setResponderInstanceUid(uid);
            const downstream = relayClientFor(this.downstream);
            if (downstream) {
                const resp = await unary(downstream.tick.bind(downstream), req, 5000);
                for (const hop of resp.getHopsList()) out.addHops(hop);
            }
            const hop = new relayPb.HopReceipt();
            hop.setSlug(slug);
            hop.setUid(uid);
            hop.setReceived(count);
            out.addHops(hop);
            callback(null, out);
        } catch (err) {
            callback(err);
        }
    }
}

function registerServer(server, options = {}) {
    server.addService(relayGrpc.RelayServiceService, new RelayService(options));
}

function relayClientFor(raw) {
    if (!raw) return null;
    if (raw.relayClient) return raw.relayClient;
    if (typeof raw.tick === 'function') return raw;
    const target = raw.target || raw.clientTarget || raw.address || raw;
    if (!target || typeof target !== 'string') return null;
    return new relayGrpc.RelayServiceClient(target, require('@grpc/grpc-js').credentials.createInsecure());
}

function responderSlug(obs) {
    const configured = String(obs && obs.cfg ? obs.cfg.slug || '' : '').trim();
    if (configured) return configured;
    return path.basename(process.argv[1] || '').replace(/\.exe$/, '');
}

function unary(method, request, timeoutMs) {
    return new Promise((resolve, reject) => {
        const timer = setTimeout(() => reject(new Error('timeout')), timeoutMs);
        method(request, (err, out) => {
            clearTimeout(timer);
            if (err) reject(err);
            else resolve(out || {});
        });
    });
}

module.exports = {
    RelayService,
    registerServer,
    RegisterServer: registerServer,
    relayPb,
    relayGrpc,
};

File: sdk/python-holons/holons/relay.py
"""Canonical RelayService implementation for cascade examples."""

from __future__ import annotations

import os
import sys
import threading
from pathlib import Path

import grpc

from holons import observability

_SDK_GEN_ROOT = Path(__file__).resolve().parents[1] / "gen" / "python"
if _SDK_GEN_ROOT.is_dir() and str(_SDK_GEN_ROOT) not in sys.path:
    sys.path.insert(0, str(_SDK_GEN_ROOT))

from relay.v1 import relay_pb2, relay_pb2_grpc  # noqa: E402


class RelayOptions:
    def __init__(self, downstream_conn: grpc.Channel | None = None) -> None:
        self.downstream_conn = downstream_conn


class _RelayService(relay_pb2_grpc.RelayServiceServicer):
    def __init__(self, downstream_conn: grpc.Channel | None = None) -> None:
        self._downstream_conn = downstream_conn
        self._received = 0
        self._lock = threading.Lock()

    def Tick(self, request, context):
        del context
        with self._lock:
            self._received += 1
            received = self._received

        obs = observability.current()
        slug = _responder_slug(obs)
        uid = obs.cfg.instance_uid
        obs.logger("tick").info(
            "tick received",
            sender=request.sender,
            note=request.note,
            responder_slug=slug,
            responder_uid=uid,
        )
        counter = obs.counter(
            "cascade_ticks_total",
            "Ticks received by this cascade node.",
            {"responder_uid": uid},
        )
        if counter is not None:
            counter.inc()

        hops = []
        if self._downstream_conn is not None:
            downstream = relay_pb2_grpc.RelayServiceStub(self._downstream_conn)
            response = downstream.Tick(request, timeout=5.0)
            hops.extend(response.hops)
        hops.append(relay_pb2.HopReceipt(slug=slug, uid=uid, received=received))
        return relay_pb2.TickResponse(
            responder_slug=slug,
            responder_instance_uid=uid,
            hops=hops,
        )


def register_server(server: grpc.Server, options: RelayOptions | None = None) -> None:
    options = options or RelayOptions()
    relay_pb2_grpc.add_RelayServiceServicer_to_server(
        _RelayService(options.downstream_conn),
        server,
    )


def _responder_slug(obs: observability.Observability) -> str:
    configured = obs.cfg.slug.strip()
    if configured:
        return configured
    if sys.argv:
        return Path(sys.argv[0]).name
    return ""


__all__ = ["RelayOptions", "register_server", "relay_pb2", "relay_pb2_grpc"]
